src/SonarScanner.MSBuild.PreProcessor/WebService/SonarQubeWebService.cs

```diff
@@ -0,0 +1,92 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using SonarScanner.MSBuild.Common;
+
+namespace SonarScanner.MSBuild.PreProcessor.WebService
+{
+    internal class SonarQubeWebService : SonarWebService
+    {
+        public SonarQubeWebService(IDownloader downloader, Uri serverUri, Version serverVersion, ILogger logger)
+            : base(downloader, serverUri, serverVersion, logger)
+        {
+            // ToDo: Fail fast after release of S4NET 6.0
+            if (serverVersion.CompareTo(new Version(7, 9)) < 0)
+            {
+                logger.LogWarning(Resources.WARN_SonarQubeDeprecated);
+            }
+        }
+
+        public override async Task<bool> IsServerLicenseValid()
+        {
+            logger.LogDebug(Resources.MSG_CheckingLicenseValidity);
+            var uri = GetUri("api/editions/is_valid_license");
+            var response = await downloader.TryGetLicenseInformation(uri);
+            var content = await response.Content.ReadAsStringAsync();
+            var json = JObject.Parse(content);
+            if (response.StatusCode == HttpStatusCode.NotFound)
+            {
+                if (json["errors"]?.Any(x => x["msg"]?.Value<string>() == "License not found") == true)
+                {
+                    return false;
+                }
+
+                logger.LogDebug(Resources.MSG_CE_Detected_LicenseValid);
+                return true;
+            }
+            else
+            {
+                return json["isValidLicense"].ToObject<bool>();
+            }
+        }
+
+        public override bool IsSonarCloud() => false;
+
+        protected override async Task<IDictionary<string, string>> DownloadComponentProperties(string projectId) =>
+            serverVersion.CompareTo(new Version(6, 3)) >= 0
+                ? await base.DownloadComponentProperties(projectId)
+                : await GetComponentPropertiesLegacy(projectId);
+
+        protected override Uri AddOrganization(Uri uri, string organization) =>
+            string.IsNullOrEmpty(organization) || serverVersion.CompareTo(new Version(6, 3)) < 0
+                ? uri
+                : new Uri(uri + $"&organization={WebUtility.UrlEncode(organization)}");
```

Comment:
Ideas: 
* IsNullOrEmpty is already checked in the base class and should not be re-checked here
* The URL building is complex and prone to errors and typos and out-of-sync issues =>should use base class
```suggestion
            serverVersion.CompareTo(new Version(6, 3)) < 0 ? uri : base.AddOrganization(uri, organization));
```

Reply:
I don't want to change this now. It can be done at a later time.

Reply:
This one is quite important. As the current version has a lot of surface for errors. And rebase should be easy, as signature isn't changed.